### PR TITLE
Notify MastodonPlugins about a Mastodon project being closed

### LINKED ISSUE #237: Notify MastodonPlugins about a Mastodon project being closed
# Problem

Currently there is no way for a MastodonPlugin to get notified when the Mastodon project is closed. This means it cannot free resources, that are associated with the Mastodon project or do other operations like storing settings etc. that would usually be done when closing a project.

This is a problem for the Lineage Registration plugin in mastodon-tomancak. The plugin keeps a list of opened Mastodon projects. But a Mastodon project is never removed from that list because the plugin simply doesn't know if a project gets closed or not.

# Potential Solution (deprecated)

<s>
The `MastodonPlugin` interface has a method that is called whenever a new Mastodon project is opened:

https://github.com/mastodon-sc/mastodon/blob/a965612e7890ec624232fc69c7431badd65c18d3/src/main/java/org/mastodon/app/plugin/MastodonPlugin.java#L51

We could add a similar method:
```java
void closingAppPluginModel( final M pluginAppModel )`
```
that is called whenever the Mastodon project is closed.</s> 

# Preferred Solution

Add `closingListerners` to the `AppModel`. A plugin can listen to the `AppModel` and would be notified when the `AppModel` is closed.

### PR DESCRIPTION
**Pull request by maarzt**
Fixes #237

This PR does two things:
* Adds a CloseListener to the MamutAppModel, this allows a MamutPlugin to get notified when the Mastodon project is closed.
* Allows a MamutPlugin to retrieve the project folder when initializing the plugin.

Here is an example:
```java
@Plugin( type = MamutPlugin.class )
public class ExampleMastodonPlugin implements MamutPlugin
{
	@Override
	public void setAppPluginModel( MamutPluginAppModel appPluginModel )
	{
		// register a close listener
		appPluginModel.getAppModel().closeListeners().add( this::onClosed );
		
		// get the project folder
		File projectFolder = appPluginModel
				.getWindowManager()
				.getProjectManager()
				.getProject()
				.getProjectRoot();
	}

	private void onClosed()
	{
		// TODO: free resources
	}
}
```

### CODE REVIEW

**Comment by tinevez** [comment]
It's good. I see that you plugged the listener into the branch views. The branch views have already a `close()` method that should be called when the main window is closed. It does not work in your cases?

File: src/main/java/org/mastodon/mamut/MamutAppModel.java [hunk 1]
```diff
@@ -157,4 +160,16 @@ public BranchGraphSynchronizer getBranchGraphSync()
...
+
+	public void close() {
+		closeListeners.list.forEach( CloseListener::close );
+	}
+
+	/**
+	 * Listeners that are notified when the Mastodon project is closed,
+	 * by for example closing the main window.
+	 */
+	public Listeners< CloseListener > closeListeners() {
```
Comment (stefanhahmann) [inline src/main/java/org/mastodon/mamut/MamutAppModel.java:172]: The method name may be misleading and could also be interpreted as in "close the listeners". The javadoc explains it, but nevertheless maybe `getCloseListeners()` would be less ambiguous?

File: src/main/java/org/mastodon/mamut/MamutAppModel.java [hunk 2]
```diff
@@ -157,4 +160,16 @@ public BranchGraphSynchronizer getBranchGraphSync()
 	{
 		return branchGraphSync;
 	}
+
+	public void close() {
```
Comment (stefanhahmann) [inline src/main/java/org/mastodon/mamut/MamutAppModel.java:164]: Could you please check, if you have the correct code style settings? The bracket for the method should be a new line.

**Review by stefanhahmann — CHANGES_REQUESTED** [review]
The PR adds the ability that Plugins can add listeners that inform, when a project is closed. This is a well needed functionality.

I left some comments re code style and method naming.

File: src/main/java/org/mastodon/mamut/MamutAppModel.java (hunk 1 above)
Comment (maarzt) [reply src/main/java/org/mastodon/mamut/MamutAppModel.java]: I renamed the method to projectCloseListeners()

**Review by stefanhahmann — APPROVED** [review]
My comments have been resolved. All good now from my perspective.

**Review by tinevez — APPROVED** [review]

**Pull request merged**